### PR TITLE
Check recipient addresses for ENS confusables in batch send

### LINKED ISSUE #530: Add check confusable to batch send and format error message to not overlap send
Screenshot from 2023-06-05 15-48-26

Change formatting so it does not overlap and also check for confusables.

### LINKED ISSUE #551: The warning about ENS Confusables is not shown for ENS names entered on the batch send page

### PR DESCRIPTION
**Pull request by jferas**
Resolves #551 
Resolves #530

### CODE REVIEW

File: frontend/src/pages/AccountSend.vue [hunk 1]
```diff
@@ -806,11 +806,14 @@ function useSendForm() {
   // Validators
   async function isValidId(val: string | undefined) {
     // Return true if nothing is provided
-    checkConfusables();
     if (!val) return true;
 
     // Check if recipient ID is valid
     try {
+      // Check if confusable chars in string, throws with warning if so
+      checkConfusables(val);
```
Comment (alexkeating) [inline frontend/src/pages/AccountSend.vue:814]: I think this change causes a confusable character to show an error on the frontend rather than a warning. This will prevent users from sending to users that have confusable characters in their ENS name. @garyghayrat is this the behavior we want?

File: frontend/src/pages/AccountSend.vue [hunk 2]
```diff
@@ -366,7 +366,7 @@
                       placeholder="vitalik.eth"
                       :label="$t('Send.receiver-addr-ens')"
                       lazy-rules
-                      :rules="isValidId"
+                      :rules="(value: string) => isValidId(value)"
```
Comment (alexkeating) [inline frontend/src/pages/AccountSend.vue:369]: Is this change necessary? Is there a different behavior when the rules are defined this way?

File: frontend/src/pages/AccountSend.vue [hunk 3]
```diff
@@ -1206,18 +1209,11 @@ function useSendForm() {
...
-      recipientIdWarning.value = undefined;
-    } catch (e) {
-      if (e instanceof Error) {
-        recipientIdWarning.value = e.message;
-      }
+  function checkConfusables(recipientIdString: string | undefined) {
+    if (recipientIdString && recipientIdString.endsWith('eth')) {
+      assertValidEnsName(recipientIdString);
     }
+    recipientIdWarning.value = undefined;
```
Comment (alexkeating) [inline frontend/src/pages/AccountSend.vue:1216]: Is this needed? Will this supress all warnings?

File: frontend/src/pages/AccountSend.vue (hunk 2 above)
Comment (jferas) [reply frontend/src/pages/AccountSend.vue:369]: > Is this change necessary? Is there a different behavior when the rules are defined this way?

I think this is necessary because we need to pass the value of the text input object being changed into the function `isValidId` (and ultimately into `checkConfusables`). If we left it the way it was, inside `checkConfusables`.. where we do the following:
```
    const recipientIdString = recipientId.value || '';
```
We end up always getting the contents of the text object on the send page but not one of the objects on the batch send page. So we end up checking confusables on the wrong string.

File: frontend/src/pages/AccountSend.vue (hunk 1 above)
Comment (jferas) [reply frontend/src/pages/AccountSend.vue:814]: > I think this change causes a confusable character to show an error on the frontend rather than a warning. This will prevent users from sending to users that have confusable characters in their ENS name. @garyghayrat is this the behavior we want?

Ah... I see... looks like I've made this too restrictive.. thanks.. I'll fix

Comment (garyghayrat) [reply frontend/src/pages/AccountSend.vue:814]: I'd say warning is sufficient.

File: frontend/src/pages/AccountSend.vue (hunk 3 above)
Comment (jferas) [reply frontend/src/pages/AccountSend.vue]: Gone: [34740c2](https://github.com/ScopeLift/umbra-protocol/pull/640/commits/34740c2777b044c67badac8536a056d80aadd2ca)

File: frontend/src/pages/AccountSend.vue (hunk 1 above)
Comment (jferas) [reply frontend/src/pages/AccountSend.vue]: Restored to warning.. with a new warning element associated with each batch send entry: [34740c2](https://github.com/ScopeLift/umbra-protocol/pull/640/commits/34740c2777b044c67badac8536a056d80aadd2ca)

File: frontend/src/pages/AccountSend.vue [hunk 4]
```diff
@@ -116,7 +116,7 @@
             placeholder="vitalik.eth"
             lazy-rules
             :hideBottomSpace="true"
-            :rules="isValidId"
+            :rules="(value: string) => isValidId(value, undefined)"
```
Comment (alexkeating) [inline frontend/src/pages/AccountSend.vue:119]: (Arbitrary thread)

- Still reviewing, but I found a couple weird scenarios

- It takes a while for the error to show up on this one.
Screenshot from 2024-01-26 09-43-34

- Some state must not be getting cleared
[Screencast from 2024-01-26 11-01-27.webm](https://github.com/ScopeLift/umbra-protocol/assets/12705146/1de2d19d-9f5b-490d-86be-dffc0e712026)

Comment (jferas) [reply frontend/src/pages/AccountSend.vue:119]: Thanks for the review! The warning state not getting cleared issue is resolve here: [f8ff428](https://github.com/ScopeLift/umbra-protocol/pull/640/commits/f8ff428ffc58a3d2219fa6b13724c43538b3e840)

The issue related to the slow error appearance on the batch send form looks like it is also present in the existing release, and appears to be related to the way that the "batchSend"s are "watched" and validations processed.
Might be a good idea to make a separate PR for this issue.

File: frontend/src/pages/AccountSend.vue [hunk 5]
```diff
@@ -803,14 +808,25 @@ function useSendForm() {
...
+    // Check if confusable chars in string, throws with warning if so
+    checkConfusables(val, index);
+
     // Return true if nothing is provided
-    checkConfusables();
     if (!val) return true;
 
     // Check if recipient ID is valid
     try {
+      // Check if ENS name is valid
```
Comment (alexkeating) [inline frontend/src/pages/AccountSend.vue:829]: I think this can be removed now

File: frontend/src/pages/AccountSend.vue [hunk 6]
```diff
@@ -1206,16 +1222,15 @@ function useSendForm() {
...
     try {
       if (recipientIdString && recipientIdString.endsWith('eth')) {
         assertValidEnsName(recipientIdString);
       }
-      recipientIdWarning.value = undefined;
-    } catch (e) {
-      if (e instanceof Error) {
-        recipientIdWarning.value = e.message;
+      setWarning('', index); // clear warning
+    } catch (e: unknown) {
```
Comment (alexkeating) [inline frontend/src/pages/AccountSend.vue:1231]: I think this `unknown` can be removed. Explicit type declarations can cause issues down the line if they are un-needded

**Review by alexkeating — COMMENTED** [review]
Looks good! A couple nits and this should be ready

**Review by alexkeating — APPROVED** [review]

File: frontend/src/pages/AccountSend.vue (hunk 4 above)
Comment (jferas) [reply frontend/src/pages/AccountSend.vue:119]: Ended up putting the fix for this in the PR after all, as it was so small: https://github.com/ScopeLift/umbra-protocol/pull/640/commits/62a4bf2f9fb0ebf688b32ff9627ab114a334713e

Comment (alexkeating) [reply frontend/src/pages/AccountSend.vue:119]: The formatting is a little weird in the batch view. This should probably be resolved if possible.

Screenshot from 2024-01-30 14-48-52
Screenshot from 2024-01-30 14-49-57
Screenshot from 2024-01-30 14-50-24

Comment (jferas) [reply frontend/src/pages/AccountSend.vue:119]: Good point.. I think the "error overlapping the ID object below" issue also exist in the current release.. but the warning issue didn't (as that's what this PR was intended to solve.. the missing warning). I should be able to get both resolved in the next commit to this PR.

Comment (jferas) [reply frontend/src/pages/AccountSend.vue:119]: Fix the overlaps on both batch send mobile and desktop..however, if the page is very narrow (just before it transitions to mobile layout) there is still an overlap of the error message onto the warning message .. this is because the base-input component is dropping down a very narrow paragraph (because of the screen width) of the long error text directly beneath it. [8459e97](https://github.com/ScopeLift/umbra-protocol/pull/640/commits/8459e97d09a7a8e2c77bce4b128362e9dfbbfd53)

Comment (garyghayrat) [reply frontend/src/pages/AccountSend.vue:119]: Sorry for the last min nits, but 
1. Could you reduce the distance between the error message and the confusables warning message on Desktop? Ideally to the same distance that they are on the Send page.



2. Could you make the fonts size of the yellow warning message same as the red error message? Maybe they already are and it's just my browser?

Comment (alexkeating) [reply frontend/src/pages/AccountSend.vue:119]: Looking good @jferas, I think after these final nits we this will be ready

Comment (jferas) [reply frontend/src/pages/AccountSend.vue:119]: Done: fixed the warning/error layout issue for desktop at various widths:https://github.com/ScopeLift/umbra-protocol/pull/640/commits/f783e6d49a57cb3e9afef52eae28e1e003264a2c
Also fixed a small issue that only showed on an actual device where error overlapped warning: https://github.com/ScopeLift/umbra-protocol/pull/640/commits/964b574f18d446ad024bf310670faf227a8cad8a

Comment (garyghayrat) [reply frontend/src/pages/AccountSend.vue:119]: A couple more nits, 
1. the x-axis padding or margins are a bit larger than the error message.
SCR-20240205-lwoc

2. On mobile view, the batch send text box distance between the error and the warning are bit larger than Single Send. 

**Single Send:**
image

**Batch Send:**
image

3. Weird behavior where if you switch between Single Send and Batch Send while putting in an unregistered address with confusables, Single Page reacts by hiding the error message, and moving the warning message up closer to the address input box. Batch Send page keeps the warning message in the same spot, with a huge margin where the error message would be. 

**Single Send**:
image

**Batch Send**:
image

Comment (jferas) [reply frontend/src/pages/AccountSend.vue:119]: Nits 1 and 2 were relatively minor.. done: [8bb038e](https://github.com/ScopeLift/umbra-protocol/pull/640/commits/8bb038e59c6075f371ddc77b9c7c57a53e162b95)

Nit 3 was kind of tricky.. the error message was automatically being removed by Vue on the tab switch because it was part of the recipient ID base-input component, but the warning message (which was generated into its own object was not getting removed on tab switch, so added logic to do that so that error and warning appeared to be handled the same way.. let me know if you think there's a better way of doing this: [462fac7](https://github.com/ScopeLift/umbra-protocol/pull/640/commits/462fac7ad460f99bc3e316367f31e72709192814)

**Review by alexkeating — APPROVED** [review]

**Review by garyghayrat — APPROVED** [review]
lgtm, thanks John!

File: frontend/src/pages/AccountSend.vue (hunk 4 above)
Comment (alexkeating) [reply frontend/src/pages/AccountSend.vue:119]: I am getting the slightest of overlaps

Screenshot from 2024-02-12 15-49-05
Screenshot from 2024-02-12 15-48-53

Comment (garyghayrat) [reply frontend/src/pages/AccountSend.vue:119]: @jferas will create an issue to track this.
